tests/functional/test_color.py

```diff
@@ -0,0 +1,36 @@
+import sys
+
+import pytest
+
+colorama = pytest.importorskip("pip._vendor.colorama")
+
+
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="colorama strips ANSI sequences from output on Windows",
```

Comment:
Is there a way to test WHEN=ALWAYS on windows?

Reply:
I think we may be able to mock out `colorama.AnsiToWin32.win32_calls`. This is what colorama's tests do.

Reply:
My last commit adds test cases for Windows, which attempt to monkey-patch `colorama.win32.SetConsoleTextAttribute`. Unfortunately, this cannot work because the tests run pip as a subprocess. And anyway, this is an implementation detail of colorama.

It may make more sense to add a unit test for `ColorizedStreamHandler` for all platforms, and to either exclude Windows from this functional test, or to simply assert that the ANSI sequence is never present on Windows. What do you think?

Reply:
Hmm, this is pretty much like skipping the test altogether, but I agree it's pretty annoying. I think skipping with a TODO or something is okay for now

Reply:
e089f7e640b5bc20856aaefd71da99f1f12d8a8b adds unit tests for Windows, recording calls to `SetConsoleTextAttribute` for various settings of the `color` parameter to ColorizedStreamHandler. So I believe we have pretty good test coverage for color output on Windows now.

One thing to note is that `--color=always` won't have an effect on Windows if the output is not a terminal. Color output on Windows relies on the win32 console interface, and that does not apply to pipes and the like (at least that is my understanding). Colorama checks this using `GetConsoleScreenBufferInfo` with the handles for `STDOUT` and `STDERR`, and will simply strip color sequences from the output if it detects that the console API is not available.

Reply:
Oh well, it breaks on Azure Pipelines. It did work in my Windows VM. (TM)

Reply:
Fixed in 826828a6fa8769c9cb6748a689aa86d84a4f58d7. Please see commit message for details 👍 😊

Reply:
Seems like it added some complexity to this PR, but it's ok. Good job on your hard work!